### PR TITLE
Automatic Shipping and Invoicing

### PR DESCRIPTION
**Pull request by fp-ps**

### CODE REVIEW

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/invalidation/impl/ShipmentScheduleInvalidateRepository.java [hunk 1]
```diff
@@ -608,34 +626,56 @@ public void releaseRecomputeMarkerOutOfTrx(final PInstanceId pinstanceId)
...
 		final CacheInvalidateMultiRequest multiRequest = CacheInvalidateMultiRequest.allRecordsForTable(I_M_ShipmentSchedule.Table_Name);
-		modelCacheInvalidationService.invalidate(multiRequest,  ModelCacheInvalidationTiming.CHANGE);
+		modelCacheInvalidationService.invalidate(multiRequest, ModelCacheInvalidationTiming.CHANGE);
 	}
-	
 
+	private void enqueueShipmentScheduleRecompute(
+			@NonNull final Properties context,
+			@Nullable final String trxName,
+			@NonNull final String chunkUUID)
```
Comment (metas-ts) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/invalidation/impl/ShipmentScheduleInvalidateRepository.java:662]: to me it's not clear what the chunkUUID is about. cosmin wrote
> we use it to make sure we are taking into consideration the same async batch ids that were pushed when the insert query ran

which sounds plausible but i still don't really get it. please make the code clearer and/or add comments and javadocs

File: backend/de.metas.async/src/main/java/de/metas/async/api/impl/AsyncBatchBL.java [hunk 2]
```diff
@@ -474,11 +475,11 @@ public void updateProcessedFromMilestones(@NonNull final AsyncBatchId asyncBatch
 	@NonNull
 	public AsyncBatchId newAsyncBatch(@NonNull final String asyncBatchType)
 	{
-		final I_C_Async_Batch asyncBatch = newAsyncBatch()
+		final I_C_Async_Batch asyncBatch = trxManager.callInNewTrx(() -> newAsyncBatch()
```
[the remark this answers is not in the document]
Comment (metas-ts) [reply backend/de.metas.async/src/main/java/de/metas/async/api/impl/AsyncBatchBL.java:478]: ok i see. please make sure to find a solution in the follow-up, such that wone thread gets at most one additional transaction, but not two or more additional transactions at the same time

**Review by metas-ts — COMMENTED** [review]
going to integrate; please take my comments and requests to the follow-up issue

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/async/UpdateInvalidShipmentSchedulesWorkpackageProcessorScheduler.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+import de.metas.async.AsyncBatchId;
+import de.metas.async.spi.WorkpackagesOnCommitSchedulerTemplate;
+import org.checkerframework.checker.nullness.qual.NonNull;
+
+import java.util.Optional;
+import java.util.Properties;
+
+public class UpdateInvalidShipmentSchedulesWorkpackageProcessorScheduler extends WorkpackagesOnCommitSchedulerTemplate<ShipmentSchedulesUpdateSchedulerRequest>
+		implements IShipmentSchedulesUpdateScheduler
```
[the remark this answers is not in the document]
Comment (fp-ps) [reply backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/async/UpdateInvalidShipmentSchedulesWorkpackageProcessorScheduler.java:33]: I was following invoice candidate flow, but you're right, `IShipmentSchedulesUpdateScheduler` can be dropped, and also `IInvoiceCandUpdateScheduler` on the same consideration . Agreeing also on the class name refactoring

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/invalidation/impl/ShipmentScheduleInvalidateRepository.java (hunk 1 above)
Comment (fp-ps) [reply backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/invalidation/impl/ShipmentScheduleInvalidateRepository.java:662]: javadocs added

**Pull request merged**